library/lib/serve.js

```diff
@@ -191,3 +196,45 @@ function reportServerError(error, req) {
   console.error(error)
   if (reportError) reportError(error, req)
 }
+
+
+async function sendHtml(req, res, template, location, { status, headers, data }) {
+  const scriptHashes = new Set()
+  const html = template({ location, data }, scriptHashes)
+  res.status(status).set(headers).send(html)
+}
+
+async function sendHtmlWithCspHeaders(req, res, template, location, { status, headers, data }) {
+  const scriptHashes = new Set()
+  const html = template({ location, data }, scriptHashes)
+
+  // make script hashes available for CSP middleware
+  res.locals.scriptHashes = Array.from(scriptHashes).map(hash => `'sha256-${hash}'`)
+  await addCspHeaders(req, res)
+
+  res.status(status).set(headers).send(html)
+}
+
+function createCspMiddleware(contentSecurityPolicy) {
+  const contentSecurityPolicyWithHashes = {
+    ...contentSecurityPolicy,
+    directives: {
+      ...contentSecurityPolicy.directives,
+      'script-src-elem': [
+        ...contentSecurityPolicy.directives['script-src-elem'],
+        (req, res) => res.locals.scriptHashes.join(' '),
+      ]
+    }
+  }
+  return helmet.contentSecurityPolicy(contentSecurityPolicyWithHashes)
+}
+
+function addCspHeaders(req, res) {
+  return new Promise((resolve, reject) => {
```

Comment:
What is the reason we need to wrap this in a Promise? And make 'sendHtmlWithCspHeaders' async?
For me it seems that the template rendering is just a bit more involved (namely collecting the extra script SHA's), but no extra async work?